test: Move RuntimeVerifier to K8sVerifier

pchaigno opened the pull request:
This pull request moves our RuntimeVerifier test to K8sVerifier, to be executed on our 4.9 VMs. This will allow us to execute it on our 4.19 and net-next VMs as well in the future.

The BPF programs are currently compiled to try and achieve the maximum complexity possible on 4.9 (where e.g. BPF NodePort is not supported). The Makefile will be extended in a subsequent pull request to include max-complexity targets for 4.19 and net-next.

In our K8s test pipelines, we can only access VMs through kubeconfig. Thus, to be able to compile and load the BPF programs on the VM, we define a new privileged Pod which mounts the bpffs and the Cilium source directory. All test commands are executed in this privileged Pod after uninstalling Cilium from the cluster.

Updates: #10798
Updates: #12258

pchaigno committed:
test/helpers: New WaitTerminatingPodsInNsWithFilter helper

WaitTerminatingPodsInNsWithFilter is similar to the existing
WaitTerminatingPodsInNs helper. It waits for a set of pods in state
terminating to be removed from a specific namespace, but only applies to
pods matching a given filter.

This commit also renames the helper functions
WaitCleanAllTerminatingPodsXXX to WaitTerminatingPodsXXX for brievity.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test/helpers: New WaitForSinglePod helper

WaitforPods unmarshals the output of 'kubectl get pods -o json' into a
PodList object. This unmarshalling fails where there's a single pod
returned.

This commit introduces a new helper function, a simplified version of
WaitforPods, to wait for a single pod.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -1391,6 +1407,29 @@ func (kub *Kubectl) waitForNPods(checkStatus checkPodStatusFunc, namespace strin
+func (kub *Kubectl) waitForSinglePod(checkStatus checkPodStatusFunc, namespace string, filter string, timeout time.Duration) error {
+	body := func() bool {
+		pod := v1.Pod{}
+		err := kub.GetPods(namespace, filter).Unmarshal(&pod)
```
For reviewers: see commit message for why I added this helper function. Curious if there's a simpler way to unmarshal the output when we have a single pod.

errordeveloper replied:
I think it would be simpler once we switch to using `k8s.io/client-go`, instead wrapping `kubectl` and parsing output.

errordeveloper replied:
I think this machinery also predates `kubectl wait`, you might want to try using it.

pchaigno replied:
> I think this machinery also predates `kubectl wait`, you might want to try using it.

I've opened #13909 for that.

errordeveloper reviewed (approved):
Thanks for doing this Paul, the more runtime tests we can move to Kubernetes the better! I think this looks good overall, but I don't have enough context to review the test itself, but looks just fine. I'd say that improving how we interact with Kubernetes API from the test is beyond the scope of this PR.

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -1281,6 +1281,22 @@ func (kub *Kubectl) WaitforPods(namespace string, filter string, timeout time.Du
+// containterStatuses equal to "ready". Returns true if all pods achieve
+// the aforementioned desired state within timeout seconds. Returns false and
+// an error if the command failed or the timeout was exceeded.
+func (kub *Kubectl) WaitForSinglePod(namespace string, filter string, timeout time.Duration) error {
```
You can consolidate the func args like so:

```
func (kub *Kubectl) WaitForSinglePod(namespace, filter string, timeout time.Duration) error {
```

christarazi commented on test/helpers/kubectl.go:
```diff
@@ -1281,6 +1281,22 @@ func (kub *Kubectl) WaitforPods(namespace string, filter string, timeout time.Du
 	return err
 }
 
+// WaitForSinglePod waits up until timeout seconds have elapsed for all pods in the
```
Just out of curiosity, what is the failure you're referring to in your commit msg?

> WaitforPods unmarshals the output of 'kubectl get pods -o json' into a
PodList object. This unmarshalling fails where there's a single pod
returned.

pchaigno replied:
I think it failed to unmarshal because the kubectl command doesn't return a JSON list, but a single item. It started working as soon as I replaced:
```diff
- podList := &v1.PodList{}
+ pod := v1.Pod{}
- err := kub.GetPods(namespace, filter).Unmarshal(podList)
+ err := kub.GetPods(namespace, filter).Unmarshal(&pod)
```

christarazi commented on test/k8sT/Verifier.go:
```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 Authors of Cilium
```
For `git blame`'s sake, have you tried `git mv test/runtime/verifier.go test/k8sT/Verifier.go`? (I understand if git is not able to consider this a rename due to the many changes that there made to this file.)

pchaigno replied:
I just tried... without success. Unless I keep the rename into a separate commit (which would break git bisect), git isn't able to keep track of the renaming. That file was almost entirely rewritten so maybe that's okay?

christarazi replied:
Yeah, it was worth a shot.

christarazi commented on test/k8sT/Verifier.go:
```diff
@@ -0,0 +1,75 @@
+	podManifest = "test-verifier.yaml"
+)
+
+var _ = Describe("K8sVerifier", func() {
```
I think it would be nice to describe (hehe :slightly_smiling_face:) what this test is doing and how it is different than other K8s tests. I feel it's important to call out that this is not a K8s test per se, but rather we are utilizing the K8s test infrastructure to run BPF verifier tests. I see that this is explained in the commit msg, but I think it would be good to also include in the src as well.

pchaigno replied:
Makes sense. I added a comment at the top to explain the test's objective and its particularity.

christarazi reviewed (commented):
Nice! I have a few questions / comments to address; fairly minor.

pchaigno committed:
test: Move RuntimeVerifier to K8sVerifier

This commit moves our RuntimeVerifier test to K8sVerifier, to be
executed on our 4.9 VMs. This will allow us to execute it on our 4.19
and net-next VMs as well in the future.

The BPF programs are currently compiled to try and achieve the maximum
complexity possible on 4.9 (where e.g. BPF NodePort is not supported).
The Makefile will be extended in a subsequent patchset to include
max-complexity targets for 4.19 and net-next.

In our K8s test pipelines, we can only access VMs through kubeconfig.
Thus, to be able to compile and load the BPF programs on the VM, we
define a new privileged Pod which mounts the bpffs and the Cilium source
directory. All test commands are executed in this privileged Pod after
uninstalling Cilium from the cluster.

Signed-off-by: Paul Chaignon <paul@cilium.io>
